test/plugin/derivation_endpoint_test.rb

```diff
@@ -0,0 +1,1113 @@
+      assert_match %r{filename=custom\.txt}, derivation_url
+    end
+
+    it "applies :disposition" do
+      derivation_url = @uploaded_file.derivation_url(:gray)
+      refute_match %r{disposition=}, derivation_url
+
+      @shrine.plugin :derivation_endpoint, disposition: "attachment"
+      derivation_url = @uploaded_file.derivation_url(:gray)
+      refute_match %r{disposition=}, derivation_url
+
+      derivation_url = @uploaded_file.derivation_url(:gray, disposition: "attachment")
+      assert_match %r{disposition=attachment}, derivation_url
+    end
+
+    it "applies :metadata" do
+      @uploaded_file.metadata.merge!("foo" => "bar", "baz" => "quux")
+
+      derivation_url = @uploaded_file.derivation_url(:gray)
+      serialized_file = URI(derivation_url).path[/\w+$/]
+      uploaded_file = @shrine::UploadedFile.urlsafe_load(serialized_file)
+      assert_equal Hash.new, uploaded_file.metadata
+
+      @shrine.plugin :derivation_endpoint, metadata: ["foo"]
+      derivation_url = @uploaded_file.derivation_url(:gray)
+      serialized_file = URI(derivation_url).path[/\w+$/]
+      uploaded_file = @shrine::UploadedFile.urlsafe_load(serialized_file)
+      assert_equal Hash["foo" => "bar"], uploaded_file.metadata
+
+      derivation_url = @uploaded_file.derivation_url(:gray, metadata: ["baz"])
+      serialized_file = URI(derivation_url).path[/\w+$/]
+      uploaded_file = @shrine::UploadedFile.urlsafe_load(serialized_file)
+      assert_equal Hash["baz" => "quux"], uploaded_file.metadata
+    end
+
+    it "escapes path and query params" do
+      derivation_url = @uploaded_file.derivation_url(:gray, "foo bar", filename: "foo bar")
+      assert_match /gray\/foo%20bar/,   URI(derivation_url).path
+      assert_match /filename=foo\+bar/, URI(derivation_url).query
+    end
+
+    it "generates signature from derivation name, args, params, and secret key" do
+      signatures = []
+
+      derivation_url = @uploaded_file.derivation_url(:foo)
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      derivation_url = @uploaded_file.derivation_url(:bar)
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      derivation_url = @uploaded_file.derivation_url(:foo, "foo")
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      derivation_url = @uploaded_file.derivation_url(:foo, "bar")
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      derivation_url = @uploaded_file.derivation_url(:foo, expires_in: 10)
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      derivation_url = @uploaded_file.derivation_url(:foo, expires_in: 20)
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      @shrine.plugin :derivation_endpoint, secret_key: "other_secret"
+      derivation_url = @uploaded_file.derivation_url(:foo)
+      signatures << derivation_url[/signature=(\w+)/, 1]
+
+      assert_equal signatures, signatures.uniq
+    end
```

Comment:
How about a test if signature generation algorithm gets changed?

Reply:
Well, it's currently not possible to change the signature generation algorithm via public API, so I don't think that test would be useful.

Reply:
I was thinking what if a dev submits a PR and happens to modify it and we miss it in code review. With a test, now the dev will have to change the tests as well and it gives us a second point to catch it. I felt this might be important if we care about signature being the same so as to not invalidate clients URL caches. You've been pretty good with the code reviews so it's less of a concern.

Reply:
Yes, making a change to `derivation_endpoint` that will invalidate all existing URLs, or change how new URLs are being generated, is definitely something I want to avoid (that's why I wanted the API to be complete on the first release). Ok, I think a test case that asserts that a specific signature was generated for a specific URL should be a good safety net.